lib/rust/metamodel/src/data_structures.rs

```diff
@@ -0,0 +1,195 @@
+use derivative::Derivative;
+use std::marker::PhantomData;
+use vecmap::*;
+
+
+
+// ===========
+// === IDs ===
+// ===========
+
+/// A globally unique identifier, with a type-tag.
+#[derive(Derivative)]
+#[derivative(Copy(bound = ""))]
+#[derivative(Clone(bound = ""))]
+#[derivative(Debug(bound = ""))]
+#[derivative(Eq(bound = ""))]
+#[derivative(PartialEq(bound = ""))]
+#[derivative(Ord(bound = ""))]
+#[derivative(PartialOrd(bound = ""))]
+#[derivative(Hash(bound = ""))]
+pub struct Id<T> {
+    value:  u32,
+    marker: PhantomData<*const T>,
+}
+
+impl<T> Id<T> {
+    /// Assign a new ID.
+    pub fn new() -> Self {
+        use std::sync::atomic;
+        static NEXT_ID: atomic::AtomicU32 = atomic::AtomicU32::new(0);
+        let value = NEXT_ID.fetch_add(1, atomic::Ordering::Relaxed);
+        let marker = Default::default();
+        Self { value, marker }
+    }
+}
+
+impl<T> Default for Id<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T> std::fmt::Display for Id<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.value)
+    }
+}
+
+
+
+// ==========================
+// === Densely-stored map ===
+// ==========================
+
+/// Densely-stored map from internally-produced keys.
+#[derive(Debug, Derivative, Clone)]
+#[derivative(Default(bound = ""))]
+pub struct VecMap<T> {
+    data: Vec<Option<T>>,
+}
+
+impl<T> VecMap<T> {
```

Comment:
I considered `OptVec` for this, but it wasn't exactly what I wanted. `OptVec` reuses keys, so that it can keep space overhead low during mixed `insert`/`remove` workloads. `VecMap` statically disallows key reuse, because space overhead won't be a problem for expected workloads, and I'm always paranoid about correctness (mutating is allowed, but once a key is unbound it stays unbound).

Reply:
Please add this info to docs there. Plus, I think this struct can be stored next to OptVec in such a case. It might be useful in other places of the codebase. What do u think?

Reply:
`metamodel` actually can't depend on `enso-data-structures`, because `enso-data-structures` depends on `enso-prelude`, and `enso-prelude` depends on `reflect` so that it can derive `Reflect` for its types, and `reflect` depends on `metamodel`.

I intend to, at some point in the far future, move the lazy-graph `Rust` metamodel to the `Reflect` crate, implement a proper `VecMap`-based `TypeGraph` for Rust in `metamodel`, and reverse the dependency so that `metamodel` depends on `reflect` and implements the conversion from `reflect` types to its types. That way `reflect` would be at the bottom of the dependency tree of everything, so that any crate can depend on `reflect` and derive the trait without a dependency cycle. Then metamodel can take arbitrary dependencies to do its work. However, this is all farther down the roadmap than any reasonable planning horizon. :smile:

Reply:
I've documented the design difference.